assets/js/blocks/collection-filters/utils.ts

```diff
@@ -0,0 +1,56 @@
+/**
+ * External dependencies
+ */
+import { ProductCollectionQuery } from '@woocommerce/blocks/product-collection/types';
+
+function taxonomyMapper( taxonomy: string ) {
+	const map: Record< string, string > = {
+		product_tag: 'tag',
+		product_cat: 'cat',
+	};
+
+	return taxonomy in map ? map[ taxonomy ] : `_unstable_tax_${ taxonomy }`;
+}
+
+export function formatQuery( query: ProductCollectionQuery ) {
+	if ( ! query ) {
+		return {};
+	}
+
+	const sharedParams: Array< keyof ProductCollectionQuery > = [
+		'exclude',
+		'offset',
+		'order',
+		'search',
+	];
+
+	const mappedParams: { key: keyof ProductCollectionQuery; map: string }[] = [
+		{ key: 'orderBy', map: 'orderby' },
+		{ key: 'pages', map: 'page' },
+		{ key: 'parents', map: 'parent' },
+		{ key: 'perPage', map: 'per_page' },
+		{ key: 'woocommerceStockStatus', map: 'stock_status' },
+		{ key: 'woocommerceOnSale', map: 'on_sale' },
+		{ key: 'woocommerceHandPickedProducts', map: 'include' },
```

Comment:
Mm, how about making this consistent with what you did in the `taxonomyMapper` above? Like:

```js
{
  orderBy: 'orderby',
  pages: 'page',
  // ...etc
}
```

Reply:
I wanted to do it but I was struggling to type it properly. I got this error:
```
Diagnostics:
1. Type '{ orderBy: string; pages: string; parents: string; perPage: string; woocommerceStockStatus: string; woocommerceOnSale: string; woocommerceHandPickedProducts: string; }' is missing the following properties from type 'Record<keyof ProductCollectionQuery, string>': author, exclude, inherit, offset, and 7 more. [2740]
```

Here is the declaration:
```
	const mappedParams: Record< keyof ProductCollectionQuery, string > = {
		orderBy: 'orderby',
		pages: 'page',
		parents: 'parent',
		perPage: 'per_page',
		woocommerceStockStatus: 'stock_status',
		woocommerceOnSale: 'on_sale',
		woocommerceHandPickedProducts: 'include',
	};
```

In addition, if I use object like this, I have to use `Object.keys` or `Object.entries` to loop over the object, those function remove the type for the key and just set it as `string` in the call back.

Do you have any suggestion?

Reply:
To fix the first issue, you'll need to do:

```ts
const mappedParams: Partial< Record< keyof ProductCollectionQuery, string > >
```

The second issue could be solved like this:

```ts
( Object.keys(map) as Array< keyof typeof map > ).map();
```

Reply:
Thanks for your suggestion, I tried your suggestion but still couldn't make the TS totally happy:

<img width="770" alt="Screen Shot 2023-10-21 at 17 32 14" src="https://github.com/woocommerce/woocommerce-blocks/assets/5423135/2ba6b70a-5fb0-4844-aa42-4a63031e57ab">

The code I updated following your suggestion: https://gist.github.com/dinhtungdu/66ea100fefd18316b8f011ae2fd6d7db


> `( Object.keys(map) as Array< keyof typeof map > ).map();`


Comparing this to the current map, I still think the current one is easier to read. Or just becuse I don't like type casting 😄 

```
		...mappedParams.map(
			...
		),
```

Reply:
> Comparing this to the current map, I still think the current one is easier to read. Or just becuse I don't like type casting 😄

Unfortunately, with the way some things in TS are designed, sometimes type casting is pretty much unavoidable. If you are concerned with code readability of this piece of code, you could refactor out the type casting so that it is hidden from the implementation. Indeed, as it is, this inline map function is a bit ugly, so I'd probably refactor it out.

As for your code complaining, it is because you are checking whether `param in query`, but not if `param in mappedParams`. So `mappedParams[ param ]` can be `undefined` and that is unassignable to a computed key of an object.

To solve this problem you should check whether that exists. For example:

```ts
(
  Object.keys( mappedParams ) as Array< keyof typeof mappedParams >
).map(
  ( param ) => {
      const key = mappedParams[ param ];
      return key && param in query && {
	[ key ]: query[ param ],
      }
   }
)
```

[TS Playground](https://www.typescriptlang.org/play?#code/PQKhCgAIUhRAPALgUwE4DsCGAbSATZAB2XQPQGMBLZAZyhGHEoFtCB7VRSAb0gEFEiVJQBGAVxQBZZIkx5MsyAF9IAM1RtmkAOQABAO5s25TczTlkwRAE9iNbQG5w4ZPHadIldClSrMFyAAFDTwxckQAYTZsbGRwyjZ0ASFRCVoecABIAEcxNGsALiCQsMjo2PjEgEU81GsnHNrrAEk8IvQxZhE0Btz8qO9XRCKAbSzM7nHMwkwAc2R2zu7UBsylLIBdBpRWbAVkAGVsMVmimhT0WYa8ShpCPesAGUxrNgki4LZQ8KiYuMQEugACK3e4vZ6vCTbOYAOUwZjOFyuWRM6AAbmgUHgAGIaZifb6IGhFERGWKYdBOdYuNwcLgkTqQCFvRAAeUIAMSNAymQA4gAlZpAyAAXh0qli8G0ABosgcACp8CIAaVFOmwt0QMvA1Nc7i4Xh8fgCBNKvwqnOBoIezIkPJsxCKtrZHMBNAaJmOzHQxMgHS6PSyNAAFsJ0ABrX6dH0AfhJZOQFKpzj1dM8g18-mQxS+ZvK-0BNXyPMwEmDHERYeRmVc5GOBErXlmIy2WS8wbQlGGkFJ0UT6EgAB8-WIYg02KpVDQZIsAysshwCKgivLTT985V0KzUEvxzu0AAhQqQVcldd-TfbpdHhozea+-3LW9oQJzBYjue3tjneW2d-nKsGmnTBUHIYNG0uaF4CLOoin5OJFwAHkgACm2lD9lhbSAAD4GkMYxTHMZBWXQA4cHfXtyUpLJQAgTJoEgbFKGwHxIEIM8iR7axIEQDtKFQFDEGMcNBIUMQaAAOiyaBpJgAB1ZjcHmdA0H2HiOzUco2H0JtIAAAzMWQAH0+jqPSClk2T9L0vSrJA1AXgACisyAdHDZBrG0VzvJFbCdCM85hIC2REHE7V6JgVztDRHA8i81zfMgRz7JeABKSAABJAvIcNgrE6caFlCLvO0ExWBA5AvMS7RmhhcKGNSoqGJs2yIsYTJ8LKsxQMOIScoOELxLjQSqxbPCjC6ojkmEcQUBoYbptSKQZDkBRMDGtsaDXMoL0tA9sGE4bKP7caCOYbqLAACQpPBAkoHLkFujj5og5tW11WkPAdbNT1zc8LUBK80DVbRMBocgvOHbQCHBxwaX1Hi-xPbbzQLRIgdQI9RSyKH5BQbQcZ0AFEFiAnMih9hCFHECu08wntAcgFLjh+G02+yAQTuG0XhZeUyREEDPkIbkxUmTIbi58EefeHNCVRzdObBJ5pcQICZEW2baCKRzIAUIRfVfThKBwZCUY3S0NbSbk-PSxK0TYSg8CpJxUy+pGYOsAYhGiIWRZ5UzjzN3bCyaNXEA9y2UG13XBFQA2QIBE3ZbzYPqiaHDIFtvz7cd53nFZjxUXOFDgwq26QPhX0+FQBzrGQ9zXlUZP-rR9APYzsUxkybRa3rSqiu0CcpxkcLB-3VBR+A0Dg21VsC64IuuGYTBCGIcuHOYeOjZNrJ4JMHd648idm52gG0-ydDUMuHDwD80WF3Ho8ijHpcRE8oq7y1nRP-C4hUFfeYz8-5GR-kVTqhEeoDWEgNfKz9sq5XOPlcK4DzpERImRWIz9EgBXIsgiaECro3Tug9J6f0iTPy8HWMQBBtQqDBpAReTh86uy4KoMQFBLSaVQMvcOTQdYBw+BxeWlp27pTFpQJuOsACEkAA6Zx5JkVAMgxAYB4EoBo1JMiLx4rQLgYpWQiAAFb-AkmDGglBZjoGcpkCYSgiqZAko4kMZdDaVwksvQg1ibE6wbvIxKvivCyPTgAMmCTwSAIxIC+I2EUAOkTonKHGI1cYjiJJeMyAY4x4QJINxoDrDxa9XGb3kfQ6utdD6N0RsQY+BTHpFOtuMTO7iV7pJ1jMDefi-KTG8j03pfS+naJEmKWp69K6RPafCSArYbE2KUaFVRIkYxsQrloQJcjQkKJmfEqZsSmjjJWVM+xNj1j9NOd5dYNjkk2NSfMRA8pMDQSaJIFpQT8gSVkI84sVzMi3MjsgD2-CmgSRQRdZAfzuSpSyKlBocyVEDjmqrHU4AgA)